typescript/sdk/src/crud/AbstractCrudModule.ts

```diff
@@ -0,0 +1,61 @@
+import { Logger } from 'pino';
+
+import { Address } from '@hyperlane-xyz/utils';
+
+import { HyperlaneFactories } from '../contracts/types.js';
+import { CoreConfig } from '../core/types.js';
+import { HookConfig } from '../hook/types.js';
+import { IsmConfig } from '../ism/types.js';
+import { ChainMetadataManager } from '../metadata/ChainMetadataManager.js';
+import { WarpRouteConfig } from '../metadata/warpRouteConfig.js';
+import {
+  Annotated,
+  ProtocolTypedProvider,
+  ProtocolTypedTransaction,
+  SupportedProtocolType,
+} from '../providers/ProviderType.js';
+import { ChainNameOrId } from '../types.js';
+
+export type CrudConfig = CoreConfig | WarpRouteConfig | HookConfig | IsmConfig;
+
+export type CrudModuleArgs<
+  TProtocol extends SupportedProtocolType,
+  TConfig extends CrudConfig,
+  TAddress extends Record<string, any>,
+> = {
+  addresses: Record<keyof TAddress, Address>;
+  chain: ChainNameOrId;
+  chainMetadataManager: ChainMetadataManager;
+  config: TConfig;
+  provider: ProtocolTypedProvider<TProtocol>['provider'];
+};
+
+export abstract class CrudModule<
+  TProtocol extends SupportedProtocolType,
+  TConfig extends CrudConfig,
+  TFactory extends HyperlaneFactories,
+> {
+  protected abstract readonly logger: Logger;
+
+  protected constructor(
+    protected readonly args: CrudModuleArgs<TProtocol, TConfig, TFactory>,
+  ) {}
+
+  public serialize(): string {
+    return JSON.stringify(this.args.addresses);
+  }
+
+  public abstract read(address: Address): Promise<TConfig>;
```

Comment:
i'm still not sure how exactly the read command knows which address to read

maybe I enforce it so instead of `Address` it's a keyof the addresses list

Reply:
similar to others, still unsure how the implementation comes together so will leave for now and update later as necessary